google/datalab/contrib/mltoolbox/_local_predict.py

```diff
@@ -158,11 +165,109 @@ def get_prediction_results(model_dir, data, headers, img_cols=None, show_image=T
+    if len(parts) != 2:
+      raise ValueError('Invalid model name for cloud prediction. Use "model.version".')
+
+    predict_results = ml.ModelVersions(parts[0]).predict(parts[1], predict_data)
+  else:
+    predict_results = _tf_predict(model_dir_or_id, predict_data)
+
   df_r = pd.DataFrame(predict_results)
   df_s = pd.DataFrame(display_data)
   df = pd.concat([df_r, df_s], axis=1)
   # Remove duplicate columns. All 'key' columns are duplicate here.
   df = df.loc[:, ~df.columns.duplicated()]
 
   return df
+
+
+def _batch_csv_reader(csv_file, n):
+  with file_io.FileIO(csv_file, 'r') as f:
+    args = [f] * n
+    return itertools.izip_longest(*args)
+
+
+def _get_output_schema(session, output_alias_map):
+  schema = []
+  for name in sorted(six.iterkeys(output_alias_map)):
+    tensor_name = output_alias_map[name]
+    dtype = session.graph.get_tensor_by_name(tensor_name).dtype
+    if dtype == tf.int32 or dtype == tf.int64:
+      schema.append({'name': name, 'type': 'INTEGER'})
+    elif dtype == tf.float32 or dtype == tf.float64:
+      schema.append({'name': name, 'type': 'FLOAT'})
+    else:
+      schema.append({'name': name, 'type': 'STRING'})
+
+  return schema
+
+
+def _format_results(output_format, output_schema, batched_results):
+  # Convert a dict of list to a list of dict.
+  batched_results = [dict(zip(batched_results, t)) for t in zip(*batched_results.values())]
+  if output_format == 'csv':
+    results = []
+    for r in batched_results:
+      values = [str(r[schema['name']]) for schema in output_schema]
+      results.append(','.join(values))
+  elif output_format == 'json':
+    # Default json encoder cannot handle numpy types.
+    class _JSONEncoder(json.JSONEncoder):
+
+      def default(self, obj):
+        if isinstance(obj, np.integer):
+            return int(obj)
+        elif isinstance(obj, np.floating):
+            return float(obj)
+        elif isinstance(obj, np.ndarray):
+            return obj.tolist()
+        else:
+            return super(_JSONEncoder, self).default(obj)
+
+    results = [json.dumps(x, cls=_JSONEncoder) for x in batched_results]
+  else:
+    raise ValueError('Unknown output_format %s' % output_format)
+
+  return results
+
+
+def local_batch_predict(model_dir, csv_file_pattern, output_dir, output_format, batch_size=None):
+  """ Batch Predict with a specified model.
+
+  It does batch prediction, saves results to output files and also creates an output
+  schema file. The output file names are input file names prepended by 'predict_results_'.
+
+  Args:
+    model_dir: The model directory containing a SavedModel (usually saved_model.pb).
+    csv_file_pattern: a pattern of csv files as batch prediction source.
+    output_dir: the path of the output directory.
+    output_format: csv or json.
+    batch_size: if not specified, 100 is used. Larger batch_size improves performance but may
+        cause more memory usage.
+  """
+  if batch_size is None:
```

Comment:
Right. It was there to avoid an if else on the caller side. But maybe Datalab also support default parameter values (such as if --batch_size is missing, args['batch_size'] is 100 instead of None. Let me check if that works.